src/hotspot/share/adlc/adlArena.cpp

```diff
@@ -147,8 +149,9 @@ void *AdlArena::Arealloc( void *old_ptr, size_t old_size, size_t new_size ) {
   }
 
   // See if we can resize in-place
-  if( (c_old+old_size == _hwm) &&       // Adjusting recent thing
-      (c_old+new_size <= _max) ) {      // Still fits where it sits
+  if( (c_old+old_size == _hwm) &&            // Adjusting recent thing
+      ((size_t)(_max-c_old) >= new_size) ) { // Still fits where it sits, safe from overflow
+
```

Comment:
I believe this would miss the case where we shrink an allocation in place and we are not at the high water mark, where `new_size <= old_size`, but where `c_old + old_size) == _hwm` does not hold.

Reply:
Oops, I managed to drop some of the code when pasting it into the github
comment. There should be an `else if` clause:
```
if ((c_old + old_size) == _hwm) {
  ...
} else if (new_size <= old_size) {
  return c_old;
}
```
I like this version better than the earlier version because it has only one
copy of the _hwm handling. It trades an additional pointer-delta in the
shrink-last-allocation case for avoiding a compare-and-branch in the
grow-last-allocation case, which I think is a good tradeoff, though likely
pretty minor.

I do not like the new saturated_add version.  I think that just makes things
(slightly) slower vs the pointer_delta compare, with no other benefit.